server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -275,14 +277,74 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
     }
 
     /**
-     * Copy all the objects in the context to the session.
-     * @param session The session
-     * @param context The context
+     * Sets up the context for the specified session with the RealmModel.
+     *
+     * @param origContext The original context to propagate
+     * @param targetSession The new target session to propagate the context to
+     */
+    public static void cloneContextRealmClientToSession(final KeycloakContext origContext, final KeycloakSession targetSession) {
+        cloneContextToSession(origContext, targetSession, false);
+    }
+
+    /**
+     * Sets up the context for the specified session with the RealmModel, clientModel and
+     * AuthenticatedSessionModel.
+     *
+     * @param origContext The original context to propagate
+     * @param targetSession The new target session to propagate the context to
      */
-    public static void propagateContext(KeycloakSession session, KeycloakContext context) {
-        session.getContext().setRealm(context.getRealm());
-        session.getContext().setClient(context.getClient());
-        session.getContext().setAuthenticationSession(context.getAuthenticationSession());
+    public static void cloneContextRealmClientSessionToSession(final KeycloakContext origContext, final KeycloakSession targetSession) {
+        cloneContextToSession(origContext, targetSession, true);
+    }
+
+    /**
+     * Sets up the context for the specified session.The original realm's context is used to
+     * determine what models need to be re-loaded using the current session. The models
+     * in the context are re-read from the new session via the IDs.
+     */
+    private static void cloneContextToSession(final KeycloakContext origContext, final KeycloakSession targetSession,
+            final boolean includeAuthenticatedSessionModel) {
+        if (origContext == null) {
+            return;
+        }
+
+        // setup realm model if necessary.
+        if (origContext.getRealm() != null) {
+            RealmModel realmModel = targetSession.realms().getRealm(origContext.getRealm().getId());
+            if (realmModel != null) {
+                targetSession.getContext().setRealm(realmModel);
+            }
+        }
+
+        // setup client model if necessary.
+        if (origContext.getClient() != null) {
+            RealmModel realmModel = targetSession.realms().getRealm(origContext.getClient().getRealm().getId());
```

Comment:
Last bit, double lookup should be avoided: value from line 313 can be reused if `origContext.getRealm().getId() == origContext.getClient().getRealm().getId()`.

Reply:
Done! I think the test failures are spurious (CI passed with the same commit in my branch: https://github.com/rmartinc/keycloak/actions/runs/5228904068).